bgpd/bgp_route.c

```diff
@@ -414,6 +414,29 @@ bgp_info_cmp (struct bgp *bgp, struct bgp_info *new, struct bgp_info *exist,
   newattre = newattr->extra;
   existattre = existattr->extra;
 
+  if (exist->sub_type == BGP_ROUTE_NORMAL
+      && new->sub_type != BGP_ROUTE_NORMAL
+      && newattr->distance != 0) {
+    struct prefix *exist_prefix = &(exist->net->p);
+    u_char bgp_distance = bgp_distance_apply(exist_prefix, exist,
+          exist_prefix->family, SAFI_UNICAST, bgp);
+    if (newattr->distance < bgp_distance) {
+      return -1;
+    } else if (bgp_distance < newattr->distance) {
+      return 1;
+    }
+  } else if (new->sub_type == BGP_ROUTE_NORMAL
+             && exist->sub_type != BGP_ROUTE_NORMAL
+             && existattr->distance != 0) {
+    struct prefix *new_prefix = &(new->net->p);
+    u_char bgp_distance = bgp_distance_apply(new_prefix, new,
+          new_prefix->family, SAFI_UNICAST, bgp);
+    if (existattr->distance < bgp_distance) {
```

Comment:
It looks like we are passing the route distance in to zebra in btp_zebra.c -- we set it around line 1373, which seems to be passed in as part of the "api" structure around 1399. The problem, as I understand it, is that when you change the distance on a static route to something lower than BGP's route, the static isn't "winning" in the rib, and the other way around as well... It feels like this must be a bug buried in the interaction between bgp and zebra, but it's not obvious to me where it is.

But adding an admin distance in the bgp bestpath is not the right solution here. If I get the chance, I'll poke around some more in the zebra code to see if I can figure out how this is supposed to work, and where a fix needs to be applied...